egret/common/lazy_ptdf_utils.py

```diff
@@ -68,41 +69,83 @@ def check_and_scale_ptdf_options(ptdf_options, baseMVA):
     if abs_ptdf_tol < 1e-12:
         print("WARNING: abs_ptdf_tol={0}, which is low enough it may cause numerical issues in the solver. Consider rasing abs_ptdf_tol.".format(abs_ptdf_tol*baseMVA))
 
+## to hold the indicies of the violations
+## in the model or block
+def add_monitored_branch_tracker(mb):
+    mb._lt_idx_monitored = list()
+    mb._gt_idx_monitored = list()
+
 ## violation checker
-def check_violations(m, PTDF, max_viol_add):
+def check_violations(mb, md, PTDF, max_viol_add, time=None):
 
-    NWV = np.array([pe.value(m.p_nw[b]) for b in PTDF.bus_iterator()])
+    NWV = np.array([pe.value(mb.p_nw[b]) for b in PTDF.bus_iterator()])
     NWV += PTDF.phi_adjust_array
 
     PFV  = np.dot(PTDF.PTDFM, NWV)
     PFV += PTDF.phase_shift_array
 
-    ## calculate the violations
-    gt_viol = np.nonzero(np.greater(PFV, PTDF.enforced_branch_limits))[0]
-    lt_viol = np.nonzero(np.less(PFV, -PTDF.enforced_branch_limits))[0]
+    ## calculate the negative of the violations (for easy sorting)
+    gt_viol_array = PTDF.enforced_branch_limits - PFV
+    lt_viol_array = PFV + PTDF.enforced_branch_limits
+
+    gt_viol = np.nonzero(gt_viol_array < 0)[0]
+    lt_viol = np.nonzero(lt_viol_array < 0)[0]
+
+    ## these will hold the violations 
+    ## we found this iteration
+    gt_viol = frozenset(gt_viol)
+    lt_viol = frozenset(lt_viol)
+
+    ## get the lines we're monitoring
+    gt_idx_monitored = mb._gt_idx_monitored
+    lt_idx_monitored = mb._lt_idx_monitored
+
+    ## get the lines for which we've found a violation that's
+    ## in the model
+    gt_viol_in_mb = gt_viol.intersection(gt_idx_monitored)
+    lt_viol_in_mb = lt_viol.intersection(lt_idx_monitored)
+
+    ## print a warning for these lines
+    ## check if the found violations are in the model and print warning
+    baseMVA = md.data['system']['baseMVA']
+    for i in lt_viol_in_mb:
+        bn = PTDF.branches_keys[i]
+        thermal_limit = PTDF.branch_limits_array[i]
+        print(_generate_flow_viol_warning('LB', mb, bn, PFV[i], -thermal_limit, baseMVA, time))
```

Comment:
I suspect we will want a verbosity guard on the violation print statements (and in general for the lazy PTDF logic).

Reply:
Good catch; I'll clean that up.

Reply:
Now using the Egret logger.